packages/hash/api/src/lib/error.ts

```diff
@@ -19,3 +19,10 @@ export class EntityTypeMismatchError extends Error {
     this.actualEntityTypeId = actualEntityTypeId;
   }
 }
+
+export class NotFoundError extends Error {
+  constructor(message: string) {
+    super(message);
+    this.name = "NotFoundError";
+  }
+}
```

Comment:
Not important for this PR, but this sort of error is a good candidate for adding explicit structured data under `cause`. We could set the constructor to type to accept a custom-typed `cause` option in the parameters object, so we can pass structured information.

e.g. as well as a message of `"Link with source entity ID = '${sourceEntityId}', target entity ID = '${targetEntityId}' and link type ID = '${linkTypeId}' not found."`, we could have a `cause` object that included `sourceEntityID`, `targetEntityId` etc

**Background**
As of TypeScript 4.8.2 you can provide an arbitrary cause along with an Error ([MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause)) - cause can be used either to include an original error if you're rethrowing one, but we could also use it to include structured data along with errors (e.g. a code). Prior to 4.8.2 cause was typed as Error so you couldn't include structured data, now it's typed as unknown.

Reply:
Interesting, I think we are currently doing something similar for the `EntityTypeMismatchError`:
```TypeScript
export class EntityTypeMismatchError extends Error {
  entityId: string;
  expectedEntityTypeId: VersionedUri;
  actualEntityTypeId: VersionedUri;

  constructor(
    entityId: string,
    expectedEntityTypeId: VersionedUri,
    actualEntityTypeId: VersionedUri,
  ) {
    super(
      `Expected entity with id "${entityId}" to be of type "${expectedEntityTypeId}" but got:  ${actualEntityTypeId}`,
    );
    this.name = "TypeMismatchError";
    this.entityId = entityId;
    this.expectedEntityTypeId = expectedEntityTypeId;
    this.actualEntityTypeId = actualEntityTypeId;
  }
}
```

If I'm understanding correctly, if we wanted the ability to attach things like the `sourceEntityId` to a not found error, we would need to introduce a custom `LinkNotFoundError`, rather than have a general purpose `NotFoundError`?

Reply:
I don't know if that's a good idea btw. Depends on how we want to use htem.

If the more specific errors inherited from `NotFoundError` you could still check `if (error instanceof NotFoundError)` if you didn't care which sub-class it was.

Reply:
Also dunno if it's better to use the native `cause` to store the metadata or not – in theory if this was ever consumed as pure JS it'd be less of a deviation, since `cause` is part of `Error`s already. But it probably doesn't matter for this.